ciborium/src/ser/mod.rs

```diff
@@ -327,18 +336,82 @@ macro_rules! end {
         #[inline]
         fn end(self) -> Result<(), Self::Error> {
             if self.ending {
-                self.encoder.0.push(Header::Break)?;
+                self.serializer.encoder.push(Header::Break)?;
+            }
+
+            Ok(())
+        }
+    };
+}
+
+/// `end_map` outputs cached keys if we're serializing in canonical mode
+macro_rules! end_map {
+    () => {
+        #[inline]
+        fn end(self) -> Result<(), Self::Error> {
+            #[cfg(feature = "std")]
+            if self.serializer.canonical {
+                // keys get sorted in lexicographical byte order
+                let keys = self.cache_keys;
+                let values = self.cache_values;
+
+                debug_assert_eq!(
+                    keys.len(), values.len(),
+                    "ciborium error: canonicalization failed, keys and values must have same length.");
+
+                let mut pairs = std::collections::BTreeMap::new();
+                for (key, value) in keys.iter().zip(values.iter()) {
+                    pairs.insert(key, value);
+                }
+
+                for (key, value) in pairs {
+                    self.serializer.encoder.write_all(&key)?;
+                    self.serializer.encoder.write_all(&value)?;
+                }
+            }
+
+            if self.ending {
+                self.serializer.encoder.push(Header::Break)?;
             }
 
             Ok(())
         }
     };
 }
 
-struct CollectionSerializer<'a, W> {
-    encoder: &'a mut Serializer<W>,
+/// An internal struct for serializing collections.
+///
+/// Not to be used externally, only exposed as part of the [Serializer] type.
+#[doc(hidden)]
+pub struct CollectionSerializer<'a, W> {
+    serializer: &'a mut Serializer<W>,
     ending: bool,
     tag: bool,
+
+    #[cfg(feature = "std")]
+    cache_keys: Vec<Vec<u8>>,
+    #[cfg(feature = "std")]
+    cache_values: Vec<Vec<u8>>,
+}
+
+impl<'a, W> CollectionSerializer<'a, W> {
+    pub fn new(serializer: &'a mut Serializer<W>, ending: bool, tag: bool) -> Self {
+        #[cfg(feature = "std")]
+        let capacity = match serializer.canonical {
+            true => 4,
+            false => 0,
+        };
```

Comment:
I should add a code comment to explain this: canonical serialization expects to use the vec cache for sorting keys. But when not canonicalizing, we can avoid the extra allocation by setting capacity = 0.